v2/tools/generator/internal/codegen/pipeline/detect_skipping_properties.go

```diff
@@ -0,0 +1,242 @@
+// different versions of ASO, or between different versions of a given resource.
+//
+// We need to design a long term solution to this problem, but in the meantime we just want to detect the problem to
+// ensure we don't inadvertently release an update to ASO with this significant hidden flaw.
+//
+// Detection works by scanning for properties that are dropped between versions of a resource. We keep track of all
+// these properties and can raise an error if a specific property appears more than once (implying there are two or more
+// sequence versions of that property).
+//
+// To illustrate, assume we have the following set of objects across versions of our mythical CRM service:
+//
+// v1: Person(GivenName, FamilyName, Address)
+// v2: Person(KnownAs, FullName, FamilyName)
+// v3: Person(KnownAs, FullName, FamilyName, Address)
+//
+// Scanning these types, we find:
+//
+// FamilyName: present in (v1, v2, v3); no issue
+// FullName: present in (v2, v3); no issue
+// KnownAs: also present in (v2, v3); no issue
+// GivenName: present in (v1); no issue
+//
+// Address: present in (v1, v2); issue to report, aborting the pipeline.
+//
+// Additional complexities:
+// - We need to handle type renaming between versions.
+// - When introduced, we will also need to handle property renaming between versions
+//
+func DetectSkippingProperties() Stage {
+	return MakeStage(
+		DetectSkippingPropertiesStageID,
+		"Detect properties that skip resource or object versions",
+		func(ctx context.Context, state *State) (*State, error) {
+			detector := NewSkippingPropertyDetector(state.Types(), state.ConversionGraph())
+
+			// Add resources and objects to the graph
+			for _, def := range state.Types() {
+				if t, ok := astmodel.AsPropertyContainer(def.Type()); ok {
+					err := detector.AddProperties(def.Name(), t)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+
+			err := detector.CheckForSkippedProperties()
+
+			return state, err
+		})
+}
+
+type SkippingPropertyDetector struct {
+	links              map[astmodel.PropertyReference]astmodel.PropertyReference // Individual links in chains of related properties
+	observedProperties *astmodel.PropertyReferenceSet                            // Set of properties we've observedProperties
+	types              astmodel.Types                                            // Set of all known types
+	conversionGraph    *storage.ConversionGraph                                  // Graph of conversions between types
+}
+
+// NewSkippingPropertyDetector creates a new graph for tracking chains of properties as they evolve through different
+// versions of a resource or object.
+// types is a set of all known types.
+// conversionGraph contains all of the conversions/transitions between versions.
+func NewSkippingPropertyDetector(types astmodel.Types, conversionGraph *storage.ConversionGraph) *SkippingPropertyDetector {
```

Comment:
does it make sense to make both the `skippingPropertyDetector` and `newSkippingPropertyDetector` unexported? It feels like they're a bit implementation details of the actual pipeline stage `DetectSkippingProperties` and it's not clear to me if there's any intent for other packages to actually be able to use them.

If there is and I just haven't seen it yet though feel free to ignore this.

Reply:
Yes, keeping these internal makes sense. My C# brain is still using `PascalCase` by default for types.